### PR TITLE
basecamps: enable flexible upgrades of the primitive field camp

### PR DESCRIPTION
**Pull request by mlangsdorf**
#### Summary
```SUMMARY: Features "basecamps: enable flexible upgrades of the primitive field camp"```

#### Purpose of change
Faction camps have to be upgraded in a specific order: you have to build the well before you build the wall but after you build the central kitchen.  You can't have 5 expansions until you've completed the wall because you can't build the 5th tent until then.  You can't build the radio tower until the very end.

Get rid of all that, and let people build whichever upgrade they like as long as it doesn't explicitly depend on a previous upgrade.  Build the well first, or the wall, or the radio tower, or all the tents.  The only real restriction is you have to build the north half of the central building before you build the south half.

#### Describe the solution
Add blueprint_provides and blueprint_requires fields to recipes, so there's a way to specify arbitrary flags that aren't tool qualities or items.

Move the primary store of upgrade information from the oter_id to the recipe.

Add code to read and store blueprint_provides and blueprint_requires.  A basecamp's cur_level becomes meaningless and gets replaced by blueprint_provides information.

Add some documentation.

#### Describe alternatives you've considered
Because there is no longer a single available upgrade for a camp, it's not obvious what the faction code should display.  It's currently displaying the next nominal upgrade, but that should probably be replaced with entries for each available upgrade.

blueprint_provides and blueprint_requires could be re-implemented as tool qualities or something similar, I guess.  They're fairly dissimilar concepts so I'd like to keep them separate.

blueprints are beginning to become a noticeable tumor on top of recipes, and it might make sense to move them into a subclass so only some recipes have to deal with them.

#### Additional context
FlexibleFactionCampsGui1
Here's a picture of the new mission selection menu, showing multiple available upgrades after building the first two tents and part of the central kitchen.

### CODE REVIEW

File: data/json/npcs/TALK_FACTION_CAMP.json [hunk 1]
```diff
@@ -109,7 +109,7 @@
...
+    "dynamic_line": "\n1.  Faction camps used to require a second NPC to act as overseer and camp manager, but that's been replaced by the bulletin board and two-way radio. \n2.  It used to be impossible to upgrade faction camps if there was a vehicle, even a cart, on the same map.  You can now upgrade camps even if there is a vehicle on the map, as long as the upgrade doesn't change the area under the vehicle. \n3.  Faction camps used to upgrade by completely redrawing the map.  Damage to the camp would be replaced when the camp was upgraded.  Now upgrades only change the actual area being upgraded, and you will have to repair damage to the camp yourself.  \n4.  There used to be a single path for upgrading the main camp.  Now, after you have completed the first tent, you will have many options for your next upgrade, and you can have different companions working on different upgrades of the main camp at the same time.",
```
Comment (Night-Pryanik) [inline data/json/npcs/TALK_FACTION_CAMP.json]:
```suggestion
    "dynamic_line": "\n1.  Faction camps used to require a second NPC to act as overseer and camp manager, but that's been replaced by the bulletin board and two-way radio.\n2.  It used to be impossible to upgrade faction camps if there was a vehicle, even a cart, on the same map.  You can now upgrade camps even if there is a vehicle on the map, as long as the upgrade doesn't change the area under the vehicle.\n3.  Faction camps used to upgrade by completely redrawing the map.  Damage to the camp would be replaced when the camp was upgraded.  Now upgrades only change the actual area being upgraded, and you will have to repair damage to the camp yourself.\n4.  There used to be a single path for upgrading the main camp.  Now, after you have completed the first tent, you will have many options for your next upgrade, and you can have different companions working on different upgrades of the main camp at the same time.",
```

File: doc/BASECAMP.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
+# Adding alternate basecamp upgrade paths
+
+This doesn't work yet.
+
+A basecamp upgrade path is a series of basecamp upgrade missions that upgrade the camp.  Upgrade missions are generally performed sequentially, but there is an option to have them branch.  Branched missions optionally can have further missions that require missions from other branches.
+
+Bascamp upgrade paths are defined by several related files:
+* The recipe JSONs that define what the material, tool, and skill requirements to perform an upgrademission and the blueprint mapgen, blueprint requirements, blueprint provides, and blueprint resources associated with each upgrade mission.
```
Comment (Night-Pryanik) [inline doc/BASECAMP.md]:
```suggestion
* The recipe JSONs that define what the material, tool, and skill requirements to perform an upgrade mission and the blueprint mapgen, blueprint requirements, blueprint provides, and blueprint resources associated with each upgrade mission.
```

**Comment by davidpwbrown** [comment]
This is brilliant, we are getting closer to freeform faction camps!

File: doc/BASECAMP.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    "components": [ [ [ "2x4", 28 ] ], [ [ "log", 16 ] ], [ [ "nail", 56 ] ], [ [ "stick", 24 ] ], [ [ "metal_tank", 1 ] ], [ [ "pipe", 1 ] ]
+    ]
+  },
+```
+
+The `"faction_base_camp_8"` upgrade mission can be performed after `"faction_base_camp_6"` and enables the trapping, hunting, fortification, and recruiting basecamp missions.  It adds a camp stove to the camp's inventory to represent the iron stove.
+
+## mapgen_update JSON
+
+These are standard mapgen_update JSON; see doc/MAPGEN.md for more details.  Each one should change a localized portion of the camp map and should, as much is possible, be independent of any other pieces of mapgen_update for the basecamp upgrade path.  Obviously, some bits are going to expand other bits, in which case their `"blueprint_requires"` should include the `"blueprint_provides"` of the previous upgrade missions.
```
Comment (Night-Pryanik) [inline doc/BASECAMP.md]:
```suggestion
These are standard mapgen_update JSON; see doc/MAPGEN.md for more details.  Each one should change a localized portion of the camp map and should, as much as possible, be independent of any other pieces of mapgen_update for the basecamp upgrade path.  Obviously, some bits are going to expand other bits, in which case their `"blueprint_requires"` should include the `"blueprint_provides"` of the previous upgrade missions.
```

File: doc/BASECAMP.md [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+The primitive field camp has the following upgrade path:
+1. `"faction_base_camp_0"` - the initial camp survey and the bulletin board.
+2. `"faction_base_camp_1"` - setting up the northeast tent
+3. `"faction_base_camp_2"` - digging the fire pit and constructing a bed
+4. `"faction_base_camp_3"` - adding the bookshelves to store stuff
+5. `"faction_base_camp_4"` - adding the second bed to the tent
+6. any of:
+`"faction_base_camp_5"`, `"faction_base_camp_7"`, `"faction_base_camp_9"`, `"faction_base_camp_15"`, `"faction_base_camp_18"` - add additional tents with beds in any order.
+or `"faction_base_camp_6"` - build the central kitchen.  `"faction_base_camp_8"`, `"faction_base_camp_10"`, `"faction_base_camp_11"`, `"faction_base_camp_16"` must be built in that sequence afterwards, though interleaved with the other optional upgrade missions after `"faction_base_camp_4"`.
+or `"faction_base_camp_12" - the central camp well.
```
Comment (Night-Pryanik) [inline doc/BASECAMP.md]:
```suggestion
or `"faction_base_camp_12"` - the central camp well.
```

**Pull request merged**